### ISSUE TITLE
l2 regularisation broken

### ISSUE THREAD

**Issue by jamescasbon** [NervanaSystems/coach#220]
Consider this code, copied from https://github.com/NervanaSystems/coach/blob/master/rl_coach/architectures/tensorflow_components/architecture.py#L125 : 

```

            # build the network
            self.get_model()
           
           ....

            # gradients ops
            self._create_gradient_ops()

            # L2 regularization
            if self.network_parameters.l2_regularization != 0:
                self.l2_regularization = [tf.add_n([tf.nn.l2_loss(v) for v in self.weights])
                                          * self.network_parameters.l2_regularization]
                tf.add_to_collection(tf.GraphKeys.REGULARIZATION_LOSSES, self.l2_regularization)
```

I think ...

1. `self.get_model` creates `self.total_loss` to include regularisation losses 
2. `self._create_gradient_ops` gets the derivative of `self.total_loss`
3. The L2 regularization is added after both of these

A small working patch is to switch the order and manually add the L2 reg to total loss

```

            # L2 regularization
            if self.network_parameters.l2_regularization != 0:
                self.l2_regularization = [tf.add_n([tf.nn.l2_loss(v) for v in self.weights])
                                          * self.network_parameters.l2_regularization]
                tf.add_to_collection(tf.GraphKeys.REGULARIZATION_LOSSES, self.l2_regularization)
                self.total_loss += self.l2_regularization   # Since L2 reg is created after total_loss 

            # gradients ops
            self._create_gradient_ops()

```

**Comment by galleibo-intel** [comment]
Thanks a lot @jamescasbon! That's a great catch!! 
There is currently no preset using L2 regularization, which might explain why this slipped away. 

Fixed in PR #230.
